src/org/sipdroid/sipua/PublishAgent.java

```diff
@@ -0,0 +1,215 @@
+
+	public PublishAgent(SipProvider sip_provider, UserAgentProfile user_profile, String username, String realm, String passwd, Context context) {
+		this.sip_provider = sip_provider;
+		this.user_profile = user_profile;
+		this.username = username;
+		this.realm = realm;
+		this.passwd = passwd;
+		this.next_nonce = null;
+		this.context = context;
+
+	}
+
+	public void publish() {
+		this.expire_time = SipStack.default_expires;
+		this.publish("open", expire_time, "");
+	}
+
+	public void publish(String status, int expireTime, String note) {
+		SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(context);
+		Boolean publish_enable_status = prefs.getBoolean("publish_enable", true);
+		if (publish_enable_status == true && (status=="open" || status=="close")) {
+			MessageDigest md = null;
+			String tupleId;
+			try {
+				md = MessageDigest.getInstance("MD5");
+				tupleId =  md.digest(user_profile.username.getBytes()).toString();
+			} catch (NoSuchAlgorithmException e) {
+				tupleId = user_profile.username;
+				e.printStackTrace();
+			}
+			String from = user_profile.username;
+			String entity = "sip:" + user_profile.username;
+			String xml =
+					"<?xml version=\"1.0\" encoding=\"UTF-8\"?>" +
+							"<presence xmlns=\"urn:ietf:params:xml:ns:pidf\"" +
+							" entity=\"" + entity + "\">" +
+							"<tuple id=\"" + tupleId + "\">" +
+							"<status>" +
+							"<basic>" + status + "</basic>" +
+							"<note>" + note + "</note>" +
+							"</status>" +
+							"</tuple>" +
+							"</presence>";
+			MessageFactory msgf = new MessageFactory();
+			Message req = msgf.createPublishRequest(sip_provider, new NameAddress(from), "presence", expireTime, "application/pidf+xml", xml);
+			TransactionClient t = new TransactionClient(sip_provider, req, this);
+			t.request();
+		}
+	}
+
+	public void unPublish() {
+		SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(context);
+		String from = user_profile.username;
+		Boolean publish_enable_status = prefs.getBoolean("publish_enable", true);
+		if (publish_enable_status == true) {
+			MessageFactory msgf = new MessageFactory();
+			Message req = msgf.createPublishRequest(sip_provider, new NameAddress(from), "presence", 0, null, null);
+			TransactionClient t = new TransactionClient(sip_provider, req, this);
+			t.request();
+		}
+	}
+
+	private boolean generateRequestWithProxyAuthorizationheader(Message resp, Message req) {
+		ProxyAuthenticateHeader pah = resp.getProxyAuthenticateHeader();
+		ProxyAuthorizationHeader ah1 = (new DigestAuthentication(
+				req.getTransactionMethod(), req.getRequestLine().getAddress()
+				.toString(), pah, null, null, username, passwd))
+				.getProxyAuthorizationHeader();
+		req.setProxyAuthorizationHeader(ah1);
+		return true;
+	}
+
+	public void onTransSuccessResponse(TransactionClient tc, Message resp) {
+		StatusLine status = resp.getStatusLine();
+		int code = status.getCode();
+		logger.fine(String.valueOf(status));

```

Comment:
please also use message templates and include more details about the function where the log is coming from:

```logger.fine("onTransSuccessResponse: status {0}: {1}", status.getCode(), status);```